### PR TITLE
Update documentation to reflect changes to visualizer class, and concomitant changes to Turtle and p5Template

### PR DESCRIPTION
**Pull request by katestange**
By submitting this PR, I am indicating to the Numberscope maintainers that I have read and understood the [contributing guidelines](https://numberscope.colorado.edu/doc/CONTRIBUTING/) and that this PR follows those guidelines to the best of my knowledge. I have also read the [pull request checklist](https://numberscope.colorado.edu/doc/doc/pull-request-checklist/) and followed the instructions therein.

<hr/>

In view of recent changes to the p5 visualizer class (parameter data structure and validation, handling aspect ratios and presketch function), this PR does the following:

1. updates the documentation `making-a-visualizer.md` to reflect these changes
2. updates the P5Template visualizer in the workbench to agree with said documentation and reflect these changes
3. cleans up the Turtle visualizer:  make use of parameter-specific validation, reorder/prioritize the parameters according to importance, add a picture to documentation, change default values for aesthetics.

This PR started out as looking over the Turtle visualizer, not to add functionality, but just to clean up and adapt to the recent visualizer class changes, but then became primarily a documentation PR about these new changes.  So it is a slightly hybrid PR, forgive me.

### CODE REVIEW

File: doc/making-a-visualizer.md [hunk 1]
```diff
@@ -210,7 +252,9 @@ will show it in an error dialog.
 
 ### How to document your visualizer
 
-The p5 Template visualizer and the Differences visualizer both follow our
+Documentation should be included in the source file between markdown tags.
+This will be scraped from the source and automatically compiled into the
+online documentation. The p5 Template visualizer demonstrates our
 documentation conventions.
```
Comment (gwhitney) [inline doc/making-a-visualizer.md:259]: Do you want to add a remark that it's the `/* md` comments that provide that demonstration, and that it also has detailed explanatory comments set off by `// ===` comments that are for the sake of providing more detail about visualizer writing, but are not needed in your own visualizers?

File: doc/making-a-visualizer.md [hunk 2]
```diff
@@ -270,9 +315,7 @@ the file containing its class definition and export module in the folder
 `src/visualizers`. When the frontscope client runs, it'll find your visualizer
 and compile it at runtime.
 
-The visualizers in `src/visualizers` will appear as options when you start
-Numberscope with `npm run dev` as usual.
-
+If running `npm run dev`, visualizers in `src/visualizers` will be available.
 As discussed [earlier](#develop-your-visualizer-on-the-workbench), visualizers
 that aren't ready for Numberscope users should go in
```
Comment (gwhitney) [inline doc/making-a-visualizer.md:320]: Do you want to add a remark highlighting the fact that in particular, the P5VisualizerTemplate visualizer that's been discussed throughout this document will only be visible if you `npm run dev:workbench`?

File: doc/making-a-visualizer.md [hunk 3]
```diff
@@ -126,31 +138,63 @@ running in, or access to its document object model (DOM), you shouldn't need
...
-    was left unset in the parameters dialog. Do a consistency check between
-    the sequence being visualized and the parameters.
 
 #### 🔩️ Show or stop the visualization; depart from a page element _(advanced)_
 
-You shouldn't need to implement `show()`, `stop()`, or `depart()`. You can
-learn about them from the [visualizer interface](behind-the-scenes.md)
+You shouldn't frequently need to implement `show()`, `stop()`, or `depart()`.
+You can learn about them from the [visualizer interface](behind-the-scenes.md)
 documentation, and from how they're implemented in the `P5Visualizer` base
```
Comment (gwhitney) [inline doc/making-a-visualizer.md:198]: did you want to review and update that behind-the-scenes document as well in this same PR? Or leave it for another PR? I am sure that it is also out of date in various ways.

File: src/visualizers-workbench/P5VisualizerTemplate.ts [hunk 4]
```diff
@@ -76,14 +104,14 @@ class P5VisualizerTemplate extends P5Visualizer(paramDesc) {
         super(seq)
```
Comment (gwhitney) [inline src/visualizers-workbench/P5VisualizerTemplate.ts:104]: Not sure how this remained here -- it's not actually good practice to define a constructor that only calls the super constructor, as that is the default behavior for a JavaScript class. I recommend taking out the constructor as long as we are overhauling this example, which then also lets you get rid of the import of the type GenericParamDescription, which is pretty esoteric.

File: src/visualizers-workbench/P5VisualizerTemplate.ts [hunk 5]
```diff
@@ -13,10 +13,21 @@
...
 // These comments get compiled into the Visualizer's user guide page.
 
-import type {GenericParamDescription, ParamValues} from '../shared/Paramable'
+// === Import statements ===
+// These import functionality that is used to implement aspects of
+// the visualizer.  See the referenced files for
+// further information on what these each do.
+//
+// Standard parameter functionality:
+import type {GenericParamDescription} from '../shared/Paramable'
```
Comment (gwhitney) [inline src/visualizers-workbench/P5VisualizerTemplate.ts:22]: You will likely be able to eliminate the import type of GenericParamDescription, see the remarks about the constructor below.

File: src/visualizers-workbench/P5VisualizerTemplate.ts [hunk 6]
```diff
@@ -13,10 +13,21 @@
...
 import {ParamType} from '../shared/ParamType'
+// ValidationStatus allows for validation checking in parameters
+import {ValidationStatus} from '@/shared/ValidationStatus'
+// Standard visualizer class and export:
+// INVALID_COLOR allows for initializing p5 color variables
 import {P5Visualizer, INVALID_COLOR} from '../visualizers/P5Visualizer'
 import {VisualizerExportModule} from '../visualizers/VisualizerInterface'
+// Standard sequence functionality:
 import type {SequenceInterface} from '../sequences/SequenceInterface'
```
Comment (gwhitney) [inline src/visualizers-workbench/P5VisualizerTemplate.ts:31]: As far as I can see, the type SequenceInterface isn't used anywhere in this file. I recommend removing this import and the new comment describing it.

File: src/visualizers-workbench/P5VisualizerTemplate.ts [hunk 7]
```diff
@@ -76,14 +104,14 @@ class P5VisualizerTemplate extends P5Visualizer(paramDesc) {
...
-
-        return status
+    async presketch() {
+        // === Asynchronous setup ===
+        // If any pre-computations must be run before the sketch is created,
+        // placing them in the `preSketch()` function will allow them
+        // to run asynchronously, i.e. without blocking the browser.
+        // The sketch will not be created until this function completes.
+        // By default, intializing the sequence occurs here.
+        super.presketch()
```
Comment (gwhitney) [inline src/visualizers-workbench/P5VisualizerTemplate.ts:114]: You should add the `await` before super.presketch(), since that is usually what one needs to do. And then I suggest putting a comment after that to the effect that a time-consuming initialization could go there, but if you don't have anything to do, there's no need to have a presketch() function at all -- no need to define one just to await super.presketch(). This "empty" one is just here as a placeholder, to give an anchor for the explanation.

File: src/visualizers/Turtle.ts [hunk 8]
```diff
@@ -84,15 +74,33 @@ positionally to the domain elements. Range and domain must be the same length.
...
+        description: 'coordinates of the point where drawing will start',
+        hideDescription: true,
+    },
 } as const
 
-// Turtle needs work
-// Throwing the same error on previous Numberscope website
 class Turtle extends P5Visualizer(paramDesc) {
     static category = 'Turtle'
     static description =
```
Comment (gwhitney) [inline src/visualizers/Turtle.ts:106]: As long as we are doing this, please remove the unnecessary constructor that just calls super, and then you can also delete the import of GenericParamDescription.

File: src/visualizers/Turtle.ts [hunk 9]
```diff
@@ -4,11 +4,16 @@ import {VisualizerExportModule} from './VisualizerInterface'
...
 
 /** md
 # Turtle Visualizer
 
-[image should go here]
+[<img
+  src="../../assets/img/Turtle/turtle-waitforit.png"
+  width=500
+  style="margin-left: 1em; margin-right: 0.5em"
+/>](../assets/img/Turtle/turtle-waitforit.png)
```
Comment (gwhitney) [inline src/visualizers/Turtle.ts:16]: Thanks so much for adding a pretty picture! :)

File: src/visualizers/Turtle.ts [hunk 10]
```diff
@@ -84,15 +74,33 @@ positionally to the domain elements. Range and domain must be the same length.
 - strokeColor: The color used for drawing the path.
      **/
     strokeColor: {
-        default: '#ff0000',
+        default: '#c98787',
         type: ParamType.COLOR,
```
Comment (gwhitney) [inline src/visualizers/Turtle.ts:78]: I am getting this new color scheme as pink on brick red -- is that what you'd like for the default turtle visualization? (I totally defer to your design judgment, just checking/verifying that's what you intend.)

**Review by gwhitney — COMMENTED** [review]
As usual, the comments are mostly just suggestions, please take them in that spirit. Note this is just a code review -- once this is complete, I do still need to pull and run your code. In this case I am not too worried about that, as there's very little actual code change, but I will still do it to make sure, before I merge the final version of the PR.

File: doc/making-a-visualizer.md [hunk 11]
```diff
@@ -52,39 +52,37 @@ visualizer based on p5.
...
 #### 💡️ Parameters _(often used)_
 
 Parameters are the user-facing structures that ask for control values when a
 visualizer is created. When someone running your visualizer sets a parameter
 in the graphical interface (UI), its value is written to a corresponding
-top-level property of the visualizer object. Conversely, if your code
-internally changes one of those top-level properties while the visualizer is
-running, you should refresh the corresponding parameter by calling
-`this.refreshParams()` so the person interacting with it can see the change.
+property of the `paramDesc` object. Conversely, if your code internally
```
[the remark this answers is not in the document]
Comment (katestange) [reply doc/making-a-visualizer.md:74]: ok, reverting

File: doc/making-a-visualizer.md [hunk 12]
```diff
@@ -126,31 +138,63 @@ running in, or access to its document object model (DOM), you shouldn't need
...
-a good place for one-time graphics operations, like painting the background of
-a static visualizer or setting colors and stroke options that won't change
-from frame to frame.
+There are two places to do pre-computation before the sketch begins looping
+through frames of the drawing. The first is `presketch()`, which runs
+asynchronously, meaning that the browser will not be blocked while this
+function completes. This is not a part of p5.js, but a part of Numberscope.
+
+If you implement `presketch()`, begin by calling `super.presketch()`, which
+will initialize the sequence. After this call, you have access to the
```
[the remark this answers is not in the document]
Comment (katestange) [reply doc/making-a-visualizer.md:164]: got it

File: doc/making-a-visualizer.md (hunk 2 above)
Comment (katestange) [reply doc/making-a-visualizer.md:320]: got it

File: doc/making-a-visualizer.md (hunk 1 above)
Comment (katestange) [reply doc/making-a-visualizer.md:259]: good point

File: doc/making-a-visualizer.md (hunk 3 above)
Comment (katestange) [reply doc/making-a-visualizer.md:198]: If we can leave it for another PR, I'd prefer that (just don't want to dive into it right now).  I'll add to the post-delft todo so we can make an issue...

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 4 above)
Comment (katestange) [reply src/visualizers-workbench/P5VisualizerTemplate.ts:104]: ok, I've done this.... but I can't test this visualizer (even on the main ui2 branch) because workbench isn't working right

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 5 above)
Comment (katestange) [reply src/visualizers-workbench/P5VisualizerTemplate.ts:22]: done

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 6 above)
Comment (katestange) [reply src/visualizers-workbench/P5VisualizerTemplate.ts:31]: done

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 7 above)
Comment (katestange) [reply src/visualizers-workbench/P5VisualizerTemplate.ts:114]: done

File: src/visualizers/Turtle.ts (hunk 8 above)
Comment (katestange) [reply src/visualizers/Turtle.ts:106]: done

**Comment by katestange** [comment]
I believe I have addressed @gwhitney 's review comments.

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 4 above)
Comment (gwhitney) [reply src/visualizers-workbench/P5VisualizerTemplate.ts]: You can test it by commenting out the "export module" line of Grid.ts. (You're just not going to be able to commit a version with that line commented out because of all of our pre-commit checks.) If you don't want to bother, rest assured I will test it before merging.

File: doc/making-a-visualizer.md (hunk 3 above)
Comment (gwhitney) [reply doc/making-a-visualizer.md:198]: Yup, fine to defer. Resolving.

File: src/visualizers/Turtle.ts (hunk 10 above)
Comment (gwhitney) [reply src/visualizers/Turtle.ts:77]: Roger that. My taste in colors is terrible, so let's go with yours. :) Resolving.

File: src/visualizers-workbench/P5VisualizerTemplate.ts (hunk 7 above)
Comment (gwhitney) [reply src/visualizers-workbench/P5VisualizerTemplate.ts]: I took the liberty of amplifying the remarks somewhat, hope that's OK. Resolving.

**Comment by gwhitney** [comment]
Looks good and the two modified visualizers run fine. Merging.
